packages/ui-react/src/components/LayoutGrid/LayoutGrid.tsx

```diff
@@ -0,0 +1,133 @@
+  data: Item[];
+  renderCell: (item: Item, tabIndex: number) => JSX.Element;
+};
+
+const LayoutGrid = <Item extends object>({ className, columnCount, data, renderCell }: Props<Item>) => {
+  const [focused, setFocused] = useState(false);
+  const [currentRowIndex, setCurrentRowIndex] = useState(0);
+  const [currentColumnIndex, setCurrentColumnIndex] = useState(0);
+
+  const rowCount = Math.ceil(data.length / columnCount);
+
+  const handleKeyDown = useEventCallback(({ key, ctrlKey }: KeyboardEvent) => {
+    if (!['ArrowLeft', 'ArrowRight', 'ArrowUp', 'ArrowDown', 'Home', 'End', 'PageUp', 'PageDown'].includes(key)) return;
+
+    const isOnFirstColumn = currentColumnIndex === 0;
+    const isOnLastColumn = currentColumnIndex === columnCount - 1;
+    const isOnFirstRow = currentRowIndex === 0;
+    const isOnLastRow = currentRowIndex === rowCount - 1;
+    const maxRightLastRow = (data.length % columnCount || columnCount) - 1; // Never go beyond last item
+    const maxRight = isOnLastRow ? maxRightLastRow : columnCount - 1;
+
+    switch (key) {
+      case 'ArrowLeft':
+        if (isOnFirstColumn && !isOnFirstRow) {
+          // Move to last of previous row
+          setCurrentColumnIndex(columnCount - 1);
+          setCurrentRowIndex((current) => Math.max(current - 1, 0));
+
+          return;
+        }
+        return setCurrentColumnIndex((current) => Math.max(current - 1, 0));
+      case 'ArrowRight':
+        if (isOnLastColumn && !isOnLastRow) {
+          // Move to first of next row
+          setCurrentColumnIndex(0);
+          setCurrentRowIndex((current) => Math.min(current + 1, rowCount - 1));
+
+          return;
+        }
+        return setCurrentColumnIndex((current) => Math.min(current + 1, maxRight));
+      case 'ArrowUp':
+        return setCurrentRowIndex((current) => Math.max(current - 1, 0));
+      case 'ArrowDown':
+        return setCurrentRowIndex((current) => Math.min(current + 1, rowCount - 1));
+      case 'Home':
+        if (ctrlKey) {
+          setCurrentRowIndex(0);
+        }
+        return setCurrentColumnIndex(0);
+      case 'End':
+        if (ctrlKey) {
+          setCurrentRowIndex(maxRight);
+          setCurrentColumnIndex(maxRightLastRow);
+
+          return;
+        }
+        return setCurrentColumnIndex(rowCount - 1);
+      case 'PageUp':
+        return setCurrentRowIndex((current) => Math.max(current - 1, 0));
+      case 'PageDown':
+        return setCurrentRowIndex((current) => Math.min(current + 1, rowCount - 1));
+      default:
+        return;
+    }
+  });
+
+  useEffect(() => {
+    if (focused) {
+      document.addEventListener('keydown', handleKeyDown);
+    }
+
+    return () => document.removeEventListener('keydown', handleKeyDown);
+  }, [focused, handleKeyDown, columnCount, rowCount]);
+
+  const gridRef = useRef<HTMLDivElement>(null);
+
+  // Set DOM focus to a focusable element within the currently focusable grid cell
+  useLayoutEffect(() => {
```

Comment:
I wonder if using `useLayoutEffect` is safe enough...
The way I imagined it, is programmatically applying focus by executing `.focus()` on the specific element below the `handleKeyDown`. This way we don't need to resort to `useLayoutEffect`

Reply:
That would work for the key events, but not when focused changes (whenever the grid itself gets focus/blur) or when breakpoints change the amount of columns (calculated in the other useEffect). 
I safely trust `useLayoutEffect` for this purpose, but I'm still willing try other approaches we could think of.

Reply:
Ok you convinced me with good arguments!